sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs

```diff
@@ -0,0 +1,258 @@
+        {
+            IList<PropertyData> allPropertyData = new List<PropertyData>();
+
+            // Collect the list of fields read/written by each property
+            foreach (var property in properties)
+            {
+                var returned = FindReturnedField(property, compilation);
+                var updated = FindFieldAssignment(property, compilation);
+                var data = new PropertyData(property, returned, updated);
+                allPropertyData.Add(data);
+            }
+            return allPropertyData;
+        }
+
+        private static FieldData? FindFieldAssignment(IPropertySymbol property, Compilation compilation)
+        {
+            if (property.SetMethod?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is AccessorDeclarationSyntax accessor &&
+                // We assume that if there are multiple field assignments in a property
+                // then they are all to the same field
+                accessor.DescendantNodes().FirstOrDefault(n => n is ExpressionStatementSyntax) is ExpressionStatementSyntax expression &&
+                expression.Expression is AssignmentExpressionSyntax assignment &&
+                assignment.IsKind(SyntaxKind.SimpleAssignmentExpression))
+            {
+                return ExtractFieldFromExpression(AccessorKind.Setter, assignment.Left, compilation);
+            }
+
+            return null;
+        }
+
+        private static FieldData? FindReturnedField(IPropertySymbol property, Compilation compilation)
+        {
+            // We don't handle properties with multiple returns that return different fields
+            if (property.GetMethod?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is AccessorDeclarationSyntax accessor &&
+                accessor.DescendantNodes().FirstOrDefault(n => n is ReturnStatementSyntax) is ReturnStatementSyntax returnStatement &&
+                returnStatement.Expression != null)
+            {
+                return ExtractFieldFromExpression(AccessorKind.Getter, returnStatement.Expression, compilation);
+            }
+            return null;
+        }
+
+        private static FieldData? ExtractFieldFromExpression(AccessorKind accessorKind,
+            ExpressionSyntax expression,
+            Compilation compilation)
+        {
+            var semanticModel = compilation.GetSemanticModel(expression.SyntaxTree);
+            if (semanticModel == null)
+            {
+                return null;
+            }
+
+            var strippedExpression = expression.RemoveParentheses();
+
+            // Check for direct field access: "foo"
+            if (strippedExpression is IdentifierNameSyntax &&
+                semanticModel.GetSymbolInfo(strippedExpression).Symbol is IFieldSymbol field)
+            {
+                return new FieldData(accessorKind, field, strippedExpression.GetLocation());
```

Comment:
`GetLocation()` is slow as far as I could tell and it would be better imo to call it only when you are about to report diagnostic.